library/std/src/sys/hermit/thread_parker.rs

```diff
@@ -0,0 +1,117 @@
+use super::abi;
+use crate::pin::Pin;
+use crate::sync::atomic::AtomicU64;
+use crate::sync::atomic::Ordering::{Acquire, Relaxed, Release};
+use crate::time::{Duration, Instant};
+
+// These values are larger than u32::MAX, so they never conflict with the thread's `pid`.
+const EMPTY: u64 = 0x1_0000_0000;
+const NOTIFIED: u64 = 0x2_0000_0000;
+
+// Note about atomic memory orderings:
+// Acquire ordering is necessary to obtain the token, as otherwise the parked thread
+// could perform memory operations visible before it was unparked. Since once set,
+// the token cannot be unset by other threads, the state can be reset with a relaxed
+// store once it has been read with acquire ordering.
+pub struct Parker {
+    state: AtomicU64,
+}
+
+impl Parker {
+    /// Construct the thread parker. The UNIX parker implementation
+    /// requires this to happen in-place.
+    pub unsafe fn new(parker: *mut Parker) {
+        parker.write(Parker { state: AtomicU64::new(EMPTY) })
+    }
+
+    // This implementation doesn't require `unsafe` and `Pin`, but other implementations do.
+    pub unsafe fn park(self: Pin<&Self>) {
+        if self.state.load(Acquire) == NOTIFIED {
+            self.state.store(EMPTY, Relaxed);
+            return;
+        }
+
+        let pid = abi::getpid();
+        abi::block_current_task();
+        if self.state.compare_exchange(EMPTY, pid as u64, Acquire, Acquire).is_ok() {
+            loop {
+                abi::yield_now();
+
+                if self.state.load(Acquire) == NOTIFIED {
+                    break;
+                }
+
+                abi::block_current_task();
+
+                if self.state.load(Acquire) == NOTIFIED {
+                    abi::wakeup_task(pid);
+                    break;
+                }
+            }
+        } else {
+            abi::wakeup_task(pid);
```

Comment:
Isn't this our own pid? I'm a little confused why park() is calling wake() on... itself?

Reply:
`block_current_task` changes the thread's state so it is not rescheduled after the next scheduler invocation (which the program controls, since Hermit is not preemptive). If the compare-exchange fails, it means state must be `NOTIFIED`, so the state needs to be reset for normal operation to continue (else calls to `yield_now` could block indefinitely).